# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1204170478 18000
#      Wed Feb 27 22:47:58 2008 -0500
# Node ID 2c4b0cbda85aa48fa98cde1028c84e4e1ff1ea02
# Parent  a2950622f070e8ff747fca0a2fe6685cda9fdb11
oct-stream.cc (do_read): stop reading if seek fails

diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,8 +1,12 @@
+2008-02-27  John W. Eaton  <jwe@octave.org>
+
+	* oct-stream.cc (do_read): Stop reading if seek fails.
+
 2008-02-26  John W. Eaton  <jwe@octave.org>
 
 	* ov-base-int.cc (octave_base_int_helper,
 	octave_base_int_helper_traits): New templates and specializations.
 	(octave_base_int_matrix<T>::convert_to_str_internal,
 	octave_base_int_matrix<T>::convert_to_str_internal): Use them.
 
 	* DLD-FUNCTIONS/rand.cc (do_rand): Pass name of calling function
diff --git a/src/oct-stream.cc b/src/oct-stream.cc
--- a/src/oct-stream.cc
+++ b/src/oct-stream.cc
@@ -3134,23 +3134,25 @@ do_read (octave_stream& strm, octave_idx
 		      dat = nda.fortran_vec ();
 		    }
 
 		  dat[count++] = tmp;
 
 		  elts_read++;
 		}
 
+	      int seek_status = 0;
+
 	      if (skip != 0 && elts_read == block_size)
 		{
-		  strm.seek (skip, SEEK_CUR);
+		  seek_status = strm.seek (skip, SEEK_CUR);
 		  elts_read = 0;
 		}
 
-	      if (is.eof ())
+	      if (is.eof () || seek_status < 0)
 		{
 		  if (nr > 0)
 		    {
 		      if (count > nr)
 			{
 			  final_nr = nr;
 			  final_nc = (count - 1) / nr + 1;
 			}
